### DISCUSSION TITLE
Include Custom Attributes in Doc.to_array()

### DISCUSSION THREAD

**Discussion by dancsalo** [explosion/spaCy#5382] [Help: Coding & Implementations] [answered]
Are custom attributes (accessible by "_" keyspace) able to be included in the `.to_array(cols)` method of Doc? A naive approach of:

```doc.to_array(["_"])```

results in a `KeyError`.

This question was asked in https://github.com/explosion/spaCy/issues/2072 but was not explicitly answered. I looks like the answer is no, that the user should append entries to the numpy around after calling `.to_array()`. Thank you for confirming!

## Which page or section is this issue related to?
https://spacy.io/api/doc#to_array

**Comment by adrianeboyd** [comment, marked as answer]
No, not easily, because it's possible to have custom attribute values that can't be represented as integers.

I think you might run into problems with `.from_array()` if you have appended values in your array (I don't think it has an option to ignore particular columns when loading), but if you manage the details for reloading the docs yourself, then it's certainly an option.

If you want to store large numbers of docs at once, have you looked at using `DocBin` with the `store_user_data` option? https://spacy.io/usage/saving-loading#docs The msgpack serialization of custom attributes should be pretty efficient for built-in python types.
